apps/builder/app/builder/shared/sync/sync-server.ts

```diff
@@ -1,87 +1,92 @@
 import { useEffect, useRef } from "react";
 import { useBeforeUnload } from "react-use";
+import { atom } from "nanostores";
 import { sync } from "immerhin";
 import type { Project } from "@webstudio-is/project";
 import type { Build } from "@webstudio-is/project-build";
 import type { AuthPermit } from "@webstudio-is/trpc-interface/index.server";
 import { restPatchPath } from "~/shared/router-utils";
-import { enqueue, dequeue, queueStatus } from "./queue";
 
 // Periodic check for new entries to group them into one job/call in sync queue.
 const NEW_ENTRIES_INTERVAL = 1000;
 
 // First attempts we will simply retry without changing the state or notifying anyone.
 const INTERVAL_RECOVERY = 2000;
 
+// After this amount of retries without success, we consider connection status error.
+const MAX_RETRY_RECOVERY = 5;
+
 // When we reached max failed attempts we will slow down the attempts interval.
 const INTERVAL_ERROR = 5000;
 
-const useRecoveryCheck = () => {
-  useEffect(() => {
-    const intervalId = setInterval(() => {
-      if (queueStatus.get() === "recovering") {
-        dequeue();
-      }
-    }, INTERVAL_RECOVERY);
-
-    return () => clearInterval(intervalId);
-  }, []);
+const pause = (timeout: number) => {
+  return new Promise((resolve) => setTimeout(resolve, timeout));
 };
 
-const useErrorCheck = () => {
-  useEffect(() => {
-    const intervalId = setInterval(() => {
-      if (queueStatus.get() === "failed") {
-        dequeue();
-      }
-    }, INTERVAL_ERROR);
+export type QueueStatus = "running" | "idle" | "recovering" | "failed";
 
-    return () => clearInterval(intervalId);
-  }, []);
-};
+export const queueStatus = atom<QueueStatus>("idle");
 
-const useNewEntriesCheck = ({
-  buildId,
-  projectId,
-  authToken,
-  authPermit,
-  version,
-}: UserSyncServerProps) => {
-  // save the initial version from loaded build
-  const lastVersion = useRef<number>(version);
-
-  useEffect(() => {
-    if (authPermit === "view") {
-      return;
+// polling is important to queue new transactions independently
+// from async iterator and batch them into single job
+const pollTransactions = async function* () {
+  while (true) {
+    const transactions = sync();
+    if (transactions.length > 0) {
+      queueStatus.set("running");
```

Comment:
Isn't that be better to put into retry as well
```
queueStatus.set("running");
queueStatus.set("idle");
```

So all statuses would be in a single place?
Anyway i have no strong opinion, looks better than setInterval logic.

Reply:
Not quite we could put the other logic before or after invkoking retry. Here is clear lifecycle.